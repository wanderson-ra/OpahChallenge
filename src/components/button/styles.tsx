import { heightPercentageToDP as height, widthPercentageToDP as width } from "react-native-responsive-screen";

import styled from "styled-components/native";

import { metrics } from "../../configurations/metrics";

export const Container = styled.TouchableOpacity`
    width: ${width(50)}px;
    height: ${height(5)}px;
    margin-top: ${height(5)}px;
    border-radius: ${height(1)}px;
    justify-content: center;
    background-color: ${(props): string => props.theme.button.primary};
    align-items: center;
`;

export const Title = styled.Text`
    color: ${(props): string => props.theme.button.font};
    font-size: ${metrics.font.small}px;
    font-weight: 700;
`;
