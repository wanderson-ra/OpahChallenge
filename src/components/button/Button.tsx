import React, { useCallback } from "react";
import { TextStyle, StyleProp, ViewStyle, TouchableOpacity, Text } from "react-native";

import { Container, Title } from "./styles";

export interface ButtonProps {
    testID?: string;
    enable: boolean;
    onPress: () => void;
    title: string;
    containerStyle?: StyleProp<ViewStyle>;
    textStyle?: StyleProp<TextStyle>;
}

export const Button: React.FC<ButtonProps> = (buttonProps) => {
    const { enable, onPress, title, textStyle, containerStyle, testID } = buttonProps;

    const executeOnPress = useCallback(() => {
        if (enable) {
            onPress();
        }
    }, [enable, onPress]);

    return (
        <Container
            disabled={!enable}
            as={TouchableOpacity}
            style={containerStyle}
            testID={testID ? testID : "button"}
            onPress={executeOnPress}
        >
            <Title as={Text} testID={"title"} style={[textStyle, { opacity: enable ? 1 : 0.5 }]}>
                {title}
            </Title>
        </Container>
    );
};
