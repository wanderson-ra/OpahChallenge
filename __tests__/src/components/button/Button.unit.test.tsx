import React from "react";
import renderer from "react-test-renderer";

import { ThemeProvider } from "styled-components/native";

import colorTheme from "../../../../src/configurations/themes/theme";

import { Button } from "../../../../src/components/button/Button";

describe("Test of Button", () => {
    it("Test snapshot", async () => {
        const title = "anyTitle";
        const onPress = jest.fn();
        const enable = false;

        const button: renderer.ReactTestRenderer = renderer.create(
            <ThemeProvider theme={colorTheme.light}>
                <Button title={title} enable={enable} onPress={onPress} />
            </ThemeProvider>
        );

        expect(button.toJSON()).toMatchSnapshot();
    });

    it("Test with not called", async () => {
        const title = "anyTitle";
        const onPress = jest.fn();
        const enable = false;

        const button: renderer.ReactTestRenderer = renderer.create(
            <ThemeProvider theme={colorTheme.light}>
                <Button title={title} enable={enable} onPress={onPress} />
            </ThemeProvider>
        );

        const rippleTouchable = button.root.findByProps({ testID: "button" });
        const titleText = button.root.findByProps({ testID: "title" });

        await renderer.act(async () => {
            rippleTouchable.props.onPress();
        });

        expect(titleText.props.children).toEqual(title);
    });

    it("Test with called", async () => {
        const title = "anyTitle";
        const onPress = jest.fn();
        const enable = true;

        const button: renderer.ReactTestRenderer = renderer.create(
            <ThemeProvider theme={colorTheme.light}>
                <Button title={title} enable={enable} onPress={onPress} />
            </ThemeProvider>
        );

        const rippleTouchable = button.root.findByProps({ testID: "button" });
        const titleText = button.root.findByProps({ testID: "title" });

        await renderer.act(async () => {
            rippleTouchable.props.onPress();
        });

        expect(titleText.props.children).toEqual(title);
    });
});
