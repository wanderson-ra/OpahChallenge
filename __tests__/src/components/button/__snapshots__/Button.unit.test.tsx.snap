// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of Button Test snapshot 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#880e4f",
      "borderRadius": 13.5,
      "height": 66.5,
      "justifyContent": "center",
      "marginTop": 66.5,
      "opacity": 1,
      "width": 375,
    }
  }
  testID="button"
>
  <Text
    style={
      Array [
        Object {
          "color": "#FFFFFF",
          "fontSize": 28,
          "fontWeight": "700",
        },
        undefined,
        Object {
          "opacity": 0.5,
        },
      ]
    }
    testID="title"
  >
    anyTitle
  </Text>
</View>
`;
